benchmark/mnist_benchmark.py

```diff
@@ -0,0 +1,668 @@
+    """
+    Save benchmark results to a JSON file for algorithm performance visualisation.
+
+    :param results: A dictionary of results structured as follows:
+                    {
+                        "algorithm_name": {
+                            "coreset_size_1": {
+                                "run_1": accuracy_value,
+                                "run_2": accuracy_value,
+                                ...
+                            },
+                            "coreset_size_2": {
+                                "run_1": accuracy_value,
+                                "run_2": accuracy_value,
+                                ...
+                            },
+                            ...
+                        },
+                        "another_algorithm_name": {
+                            "coreset_size_1": {
+                                "run_1": accuracy_value,
+                                "run_2": accuracy_value,
+                                ...
+                            },
+                            ...
+                        },
+                        ...
+                    }
+                    Each algorithm contains coreset sizes as keys, with values being
+                    dictionaries of accuracy results from different runs.
+    """
+    base_dir = os.path.dirname(os.path.abspath(__file__))
+    with open(
+        os.path.join(base_dir, "mnist_benchmark_results.json"), "w", encoding="utf-8"
+    ) as f:
+        json.dump(results, f, indent=4)
+
+    print("Data has been saved to 'benchmark_results.json'")
+
+
+def main() -> None:
+    """
+    Perform the benchmark for multiple solvers, coreset sizes, and random seeds.
+
+    The function follows these steps:
+    1. Prepare and load the MNIST datasets (training and test).
+    2. Perform dimensionality reduction on the training data using PCA.
+    3. Initialise solvers for data reduction.
+    4. For each solver and coreset size, reduce the dataset and train the model
+       on the reduced set.
+    5. Train the model and evaluate its performance on the test set.
+    6. Save the results, which include test accuracy for each solver and coreset size.
+    """
+    (train_data_jax, train_targets_jax, test_data_jax, test_targets_jax) = (
+        prepare_datasets()
+    )
+
+    train_data_pca = pca(train_data_jax)
+
+    results = {}
+
+    config = {
+        "epochs": 100,
+        "batch_size": 8,
+        "learning_rate": 1e-3,
+        "weight_decay": 1e-5,
+        "patience": 5,
+        "min_delta": 0.001,
+    }
+
+    # Run the experiment with 5 different random keys
+    for i in range(5):
+        key = jax.random.PRNGKey(i)
+        solvers = initialise_solvers(train_data_pca, key)
+        for getter in solvers:
+            for size in [25, 50, 100, 500, 1000, 5000]:
+                name, solver = getter(size)
+                subset, _ = solver.reduce(Data(train_data_pca))
```

Comment:
so, (see comment above) this can become `subset, _ = solver.reduce(train_data_pca)`.

Also, consider changing the variable name from `subset` to `coreset`, unless there's a good reason we're not calling it a 'coreset'. Then...